src/jit/MemoryInl.h

```diff
@@ -1390,15 +1391,70 @@ static void emitAtomic(sljit_compiler* compiler, Instruction* instr)
         sljit_s32 baseReg = SLJIT_EXTRACT_REG(addr.memArg.arg);
         sljit_s32 tmpReg = srcReg;
 
+        JITArg memValue(operands + 0);
+        sljit_s32 memValueReg = SLJIT_EXTRACT_REG(memValue.arg);
+        sljit_s32 maskReg;
+        sljit_s32 shiftReg;
+
+        if (noShortAtomic && size <= 2) {
+            operationSize = SLJIT_MOV32;
+
+            shiftReg = instr->requiredReg(2);
+            maskReg = instr->requiredReg(3);
+        }
+
+        if (SLJIT_IS_IMM(memValueReg)) {
+            return;
+        }
+
+        if (noShortAtomic && size <= 2) {
+            sljit_emit_op2(compiler, SLJIT_AND32, shiftReg, 0, baseReg, 0, SLJIT_IMM, 0x3);
+            sljit_emit_op2(compiler, SLJIT_SHL32, shiftReg, 0, shiftReg, 0, SLJIT_IMM, 3); // multiply by 8
+
+            sljit_emit_op2(compiler, SLJIT_SHL32, maskReg, 0, SLJIT_IMM, (0xffffffff) >> ((4 - size) * 8), shiftReg, 0);
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+            sljit_emit_op2(compiler, SLJIT_AND32, baseReg, 0, baseReg, 0, SLJIT_IMM, ~0x3);
+#else /* !SLJIT_32BIT_ARCHITECTURE */
+            sljit_emit_op2(compiler, SLJIT_AND, baseReg, 0, baseReg, 0, SLJIT_IMM, ~0x3);
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+            sljit_emit_op2(compiler, SLJIT_AND32, srcReg, 0, srcReg, 0, maskReg, 0);
+        }
+
         sljit_emit_atomic_load(compiler, operationSize, SLJIT_TMP_DEST_REG, baseReg);
 
+        if (noShortAtomic && size <= 2) {
+            sljit_emit_op1(compiler, SLJIT_MOV32, memValueReg, 0, SLJIT_TMP_DEST_REG, 0);
+        }
+
         if (operation != OP_XCHG) {
             tmpReg = instr->requiredReg(1);
+        }
+
+        if (noShortAtomic && size <= 2) {
+            sljit_emit_op2(compiler, SLJIT_XOR32, maskReg, 0, maskReg, 0, SLJIT_IMM, -1);
+            sljit_emit_op2(compiler, SLJIT_AND32, memValueReg, 0, memValueReg, 0, maskReg, 0);
+            sljit_emit_op2(compiler, SLJIT_XOR32, maskReg, 0, maskReg, 0, SLJIT_IMM, -1);
+        }
+
+        if (operation != OP_XCHG) {
             sljit_emit_op2(compiler, operation, tmpReg, 0, SLJIT_TMP_DEST_REG, 0, srcReg, 0);
         }
 
+        if (noShortAtomic && size <= 2) {
+            sljit_emit_op2(compiler, SLJIT_OR32, tmpReg, 0, tmpReg, 0, memValueReg, 0);
+            sljit_emit_op1(compiler, SLJIT_MOV32, memValueReg, 0, SLJIT_TMP_DEST_REG, 0);
+        }
         sljit_emit_atomic_store(compiler, operationSize | SLJIT_SET_ATOMIC_STORED, tmpReg, baseReg, SLJIT_TMP_DEST_REG);
+        if (noShortAtomic && size <= 2) {
+            sljit_emit_op1(compiler, SLJIT_MOV32, SLJIT_TMP_DEST_REG, 0, memValueReg, 0);
+        }
         sljit_set_label(sljit_emit_jump(compiler, SLJIT_ATOMIC_NOT_STORED), restartOnFailure);
+
+        if (noShortAtomic && size <= 2) {
+            sljit_emit_op2(compiler, SLJIT_AND32, tmpReg, 0, SLJIT_TMP_DEST_REG, 0, maskReg, 0);
+            sljit_emit_op2(compiler, SLJIT_LSHR32, SLJIT_TMP_DEST_REG, 0, tmpReg, 0, shiftReg, 0);
```

Comment:
If these operations are reserved, and used and immedate, then maskReg can be reused, if that helps in some code above.